File: projeto/frontend/app/app_vox/lib/services/issue_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_client.dart';
import '../models/issue.dart';

class IssueService {
  Future<List<IssueReport>> getIssues({int? page, int? size}) async {
    final params = <String>[];
    if (page != null) params.add('page=$page');
    if (size != null) params.add('size=$size');
    final query = params.isNotEmpty ? '?${params.join('&')}' : '';
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/issues$query'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => IssueReport.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<IssueReport> getIssueById(int id) async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/issues/$id'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    return IssueReport.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<List<IssueReport>> getMyIssues() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/issues/my'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => IssueReport.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<IssueReport> createIssue(
    Map<String, String> fields, {
    List<http.MultipartFile> files = const [],
  }) async {
    final streamed = await ApiClient.multipartRequest(
      'POST',
      '${ApiClient.baseUrl}/api/issues',
      fields,
      files: files,
    );
    final response = await http.Response.fromStream(streamed);
    ApiClient.checkResponse(response);
    // Backend responde 201 Created só com header Location, sem corpo.
    if (response.body.trim().isEmpty) {
      final id = ApiClient.locationId(response);
      if (id != null) return getIssueById(id);
      // Sem header Location: busca a ocorrência mais recente do autor.
      final mine = await getMyIssues();
      if (mine.isNotEmpty) {
        mine.sort((a, b) => b.id.compareTo(a.id));
        return mine.first;
      }
      throw ApiException(
        'Ocorrência criada, mas não foi possível carregar os dados.',
      );
    }
    return IssueReport.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<void> updateIssue(
    int id,
    Map<String, String> fields, {
    List<http.MultipartFile> files = const [],
  }) async {
    final streamed = await ApiClient.multipartRequest(
      'PUT',
      '${ApiClient.baseUrl}/api/issues/$id',
      fields,
      files: files,
    );
    final response = await http.Response.fromStream(streamed);
    ApiClient.checkResponse(response);
  }

  Future<void> deleteIssue(int id) async {
    final response = await http.delete(
      Uri.parse('${ApiClient.baseUrl}/api/issues/$id'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  Future<List<IssueStatusHistory>> getIssueHistory(int id) async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/issues/$id/history'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => IssueStatusHistory.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<IssueImage>> getIssueImages(int id) async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/issues/$id/images'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => IssueImage.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> addIssueImage(
    int id, {
    required List<http.MultipartFile> files,
  }) async {
    final streamed = await ApiClient.multipartRequest(
      'POST',
      '${ApiClient.baseUrl}/api/issues/$id/images',
      const {},
      files: files,
    );
    final response = await http.Response.fromStream(streamed);
    ApiClient.checkResponse(response);
  }

  Future<void> deleteIssueImage(int id, int imageId) async {
    final response = await http.delete(
      Uri.parse('${ApiClient.baseUrl}/api/issues/$id/images/$imageId'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  Future<List<IssueReport>> getPendingIssues({int? page, int? size}) async {
    final params = <String>[];
    if (page != null) params.add('page=$page');
    if (size != null) params.add('size=$size');
    final query = params.isNotEmpty ? '?${params.join('&')}' : '';
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/moderation/issues/pending$query'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => IssueReport.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> approveIssue(int id) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/api/moderation/issues/$id/approve'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  Future<void> rejectIssue(int id) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/api/moderation/issues/$id/reject'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  /// Atualiza a ocorrência enviando o objeto completo (JSON), conforme API.md.
  Future<void> updateIssueJson(IssueReport issue, {int? councilorId}) async {
    final response = await http.put(
      Uri.parse('${ApiClient.baseUrl}/api/issues/${issue.id}'),
      headers: await ApiClient.authHeaders(),
      body: jsonEncode({
        'id': issue.id,
        'title': issue.title,
        'description': issue.description,
        'municipalityId': issue.municipalityId,
        'categoryId': issue.categoryId,
        'status': issue.status,
        'authorId': issue.authorId,
        'councilorId': councilorId,
        'neighborhood': issue.neighborhood,
        'street': issue.street,
        'number': issue.number,
        'latitude': issue.latitude,
        'longitude': issue.longitude,
      }),
    );
    ApiClient.checkResponse(response);
  }

  /// Vincula/atribui a ocorrência a um vereador (ou remove com null).
  Future<void> assignCouncilor(IssueReport issue, int? councilorId) =>
      updateIssueJson(issue, councilorId: councilorId);

  /// Atualiza apenas o status via endpoint de moderação.
  Future<void> updateIssueStatus(int id, String status, {String? note}) async {
    final payload = <String, dynamic>{'status': status};
    if (note != null) payload['note'] = note;
    final response = await http.patch(
      Uri.parse('${ApiClient.baseUrl}/api/moderation/issues/$id/status'),
      headers: await ApiClient.authHeaders(),
      body: jsonEncode(payload),
    );
    ApiClient.checkResponse(response);
  }
}

File: projeto/frontend/site/vox/src/app/services/issue.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface IssueReport {
  id: number;
  title: string;
  description: string;
  municipalityId: number;
  categoryId: number;
  status: string;
  authorId: number;
  createdAt: string;
  updatedAt: string;
  neighborhood: string;
  street: string;
  number: string;
  latitude: number;
  longitude: number;
}

export interface IssueImage {
  id: number;
  issueId: number;
  url: string;
}

export interface IssueStatusHistory {
  id: number;
  issueId: number;
  status: string;
  changedAt: string;
}

@Injectable({ providedIn: 'root' })
export class IssueService {
  private readonly API_URL = 'http://localhost:8080';

  constructor(private http: HttpClient) {}

  getIssues(page?: number, size?: number): Observable<IssueReport[]> {
    const params: string[] = [];
    if (page !== undefined) params.push(`page=${page}`);
    if (size !== undefined) params.push(`size=${size}`);
    const query = params.length ? `?${params.join('&')}` : '';
    return this.http.get<IssueReport[]>(`${this.API_URL}/api/issues${query}`);
  }

  getIssueById(id: number): Observable<IssueReport> {
    return this.http.get<IssueReport>(`${this.API_URL}/api/issues/${id}`);
  }

  getMyIssues(): Observable<IssueReport[]> {
    return this.http.get<IssueReport[]>(`${this.API_URL}/api/issues/my`);
  }

  createIssue(formData: FormData): Observable<IssueReport> {
    return this.http.post<IssueReport>(`${this.API_URL}/api/issues`, formData);
  }

  updateIssue(id: number, formData: FormData): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/api/issues/${id}`, formData);
  }

  deleteIssue(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/issues/${id}`);
  }

  getIssueHistory(id: number): Observable<IssueStatusHistory[]> {
    return this.http.get<IssueStatusHistory[]>(`${this.API_URL}/api/issues/${id}/history`);
  }

  getIssueImages(id: number): Observable<IssueImage[]> {
    return this.http.get<IssueImage[]>(`${this.API_URL}/api/issues/${id}/images`);
  }

  addIssueImage(id: number, formData: FormData): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/issues/${id}/images`, formData);
  }

  deleteIssueImage(id: number, imageId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/issues/${id}/images/${imageId}`);
  }

  getPendingIssues(page?: number, size?: number): Observable<IssueReport[]> {
    const params: string[] = [];
    if (page !== undefined) params.push(`page=${page}`);
    if (size !== undefined) params.push(`size=${size}`);
    const query = params.length ? `?${params.join('&')}` : '';
    return this.http.get<IssueReport[]>(`${this.API_URL}/api/moderation/issues/pending${query}`);
  }

  approveIssue(id: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/moderation/issues/${id}/approve`, null);
  }

  rejectIssue(id: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/moderation/issues/${id}/reject`, null);
  }
}
